File: combinationSum2/combinationSum2.cpp
#include "../include.hpp"

void print_vector(const vector<int> &board)
{
	for (const auto num : board)
	{
		cout << num << ", ";
	}
	cout << endl;
}

void print_dvector(const vector<vector<int>> &board)
{
	for (const auto row : board)
	{
		print_vector(row);
	}
}

class Solution
{
	void backtrack(const vector<int> &candidates, int target, int start, vector<int> &combination, vector<vector<int>> &res)
	{
		if (target == 0)
		{
			res.push_back(combination);
			return;
		}
		for (int i = start; i < candidates.size(); ++i)
		{
			if (i > start && candidates[i] == candidates[i - 1])
				continue ;
			if (candidates[i] > target)
				break ;
			combination.push_back(candidates[i]);
			backtrack(candidates, target - candidates[i], i + 1, combination, res);
			combination.pop_back();
		}
	}

public:
	vector<vector<int>> combinationSum2(vector<int> &candidates, int target)
	{
		vector<vector<int>> res;
		vector<int> combination;

		sort(candidates.begin(), candidates.end());
		backtrack(candidates, target, 0, combination, res);
		return (res);
	}
};

int main()
{
	Solution ss;
	vector<int> candidates = {2,5,2,1,2};
	print_dvector(ss.combinationSum2(candidates, 5));

	return (0);
}

File: combinationSum2/combinationSum2.py

from typing import List, Optional
import collections


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        curr = []
        candidates.sort()

        def dfs(i, curr, total):
            if total == target:
                res.append(curr.copy())
                return
            if i >= len(candidates) or total > target:
                return
            curr.append(candidates[i])
            dfs(i + 1, curr, total + candidates[i])
            curr.pop()

            while i + 1 < len(candidates) and candidates[i] == candidates[i + 1]:
                i += 1
            dfs(i + 1, curr, total)

        dfs(0, curr, 0)
        return res



nums, target = [1,1,1,1,2,2], 4
nums1, target1 = [2,5,2,1,2], 5


ss = Solution()


print(ss.combinationSum2(nums, target))
print(ss.combinationSum2(nums1, target1))




